<a href="https://colab.research.google.com/github/rsguevaraa/intro-met-comp/blob/main/Sesion3_Tecnicas_de_recojo_web.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Parte 1: Entendiendo el ambiente de Google Colab

Google Colab es un software que permite utilizar secciones de código (bloques o chunks) y texto en un mismo lugar. A este tipo de estructuras se les conoce como "Notebooks" (Cuadernos) pues permiten escribir anotaciones detalladas y ordenadas a un código.




Explicación de Texto

Por ejemplo esta es una sección de texto. Aquí podemos escribir como normalmente lo hacemos en un MS Word o un Google Docs, pero también nos permite incorporar Markdown para  modificar el contenido del texto. Por ejemplo, **podemos indicar que esta sección del texto se encuentre en negrita**. O que *que este texto se encuentre en cursiva*.

Si ustedes le dan doble-click a esta sección, podrán observar que la pantalla (cuadrado) se divide en dos partes:
1. Izquierda: Texto con código
2. Derecha: Texto final (texto renderizado)

Por ejemplo, en el lado izquierdo se puede ver el texto con diferentes simbolos en texto que permiten modificarlo, como los asteriscos para especificar cuando el texto debe visualizarse en **negrita** o en *cursiva*.

Si se percatan, los espacios entre párrafos no son muy prominentes, por ello, se puede utilizar código de HTML para generar espacios entre párrafos más claros. Por ejemplo: <br> <br>

Aquí pueden observar que la diferencias entre párrafos es mayor. Esto se debe a que he incorporado el siguiente código de html"< br> < br>". O también, si observan el título, este se encuentra en formato HTML con un encabezado nivel 2.

En el lado derecho de la pantalla, lo que se visualiza es el texto "renderizado" eso quiere decir, se visualiza el texto luego de haber aplicado el lenguaje Markdown. Si ustedes quieren dejar de ver esta pantalla dividida en dos, deberán dar click a cualquiera de los siguiente bloques, ya sea de texto o de código.

In [ ]:
# Esta es una sección de código y permite también incluir texto como este (anotaciones).
# Las anotaciones en el código son de color verde y deben ser precedidas por un #

import pandas as pd # este código sirve para importar la libreria "pandas" que permite modificar bases de datos


Una de las ventajas de Google Colab es que permite trabajar con código de Python y R al mismo tiempo. En este sentido, si queremos utilizar código de R, primero debemos instalar la lbreria "rpy2" en nuestro Google Colab. Para ello se deben seguir los siguientes dos pasos:

In [ ]:
# Paso 1: Con el comando !pip se instala la libreria "rpy2"
!pip install rpy2

# En caso ya tengan esta librería instalada no es necesario este paso.
# Por otro lado, la instalación de librerias es por única vez.
# Por ende, cuando vuelvan a abrir el Colab en su dispositivo, no es necesario volver a instalar las librerias.

In [ ]:
# Paso 2: Una vez que tengamos la libreria "rpy2" tenemos que cargar este paquete utilizando el siguiente código
%load_ext rpy2.ipython

# Al costado izquierdo veremos que aparece un check verde y un número. Eso indica que el código se ejecutó sin problemas y el tiempo que le tomo (2segundos)

Si queremos utilizar código de R, nuestro "bloque de código" (chunk) siempre debe iniciar con el siguiente comando:
```
%%R
```
Este comando le indica al programa que en esta sección de código se está utilizando el lenguaje R y no Python.

Esto es importante ya que de no especificar este detalle, el programa asumirá que el código es Python y generará errores.

In [ ]:
%%R #Siempre que queramos utilizar código en R.

# A continuación utilizaremos la librería "dplyr" y traeremos a nuestro ambiente de trabajo la base de datos que contiene personajes de star wars
library(dplyr)
sw <- starwars #asignar la base "starwars" al objeto "sw"
head(sw) #obtener las primeras filas de la base de datos especificada


# A tibble: 6 × 14
  name      height  mass hair_color skin_color eye_color birth_year sex   gender
  <chr>      <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
1 Luke Sky…    172    77 blond      fair       blue            19   male  mascu…
2 C-3PO        167    75 <NA>       gold       yellow         112   none  mascu…
3 R2-D2         96    32 <NA>       white, bl… red             33   none  mascu…
4 Darth Va…    202   136 none       white      yellow          41.9 male  mascu…
5 Leia Org…    150    49 brown      light      brown           19   fema… femin…
6 Owen Lars    178   120 brown, gr… light      blue            52   male  mascu…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


# Parte 2: Extraer información web usando la API de Wikipedia

El objetivo es extraer toda la información de la página web  ["Época del Terrorismo en el Perú"](https://es.wikipedia.org/wiki/%C3%89poca_del_terrorismo_en_el_Per%C3%BA) para entender de qué manera los conceptos de "terrorismo" y "conflicto armado interno" están siendo utilizados por los editores de la página. Para ello, realizaremos los siguientes pasos:

1.   Entender la estructura y los diferentes parámetros de la API de Wikipedia (Wikimedia).
2.   Guardar la infomación en un formato estructurado (JSON file).




## Parte 2.1: Extracción del contenido de la página (demostración de código)

In [ ]:
# Primero importaremos la libreria "requests" que nos ayudará a conectar con la API de Wikipedia
import requests

#Check verde al costado indica que el código se ejecutó sin problemas

La lógica es la siguiente:

1.   Usar la función "get" para conectar con la API de Wikipedia. Especificando el enlace (URL) y los diferentes parámetros requeridos.
2.   Almacenar la respuesta del servidor en un objeto llamado "respuesta".
3. Visualizar la respuesta utilizando el comando .json




In [ ]:
# Primero crearemos el objeto "respuesta" que almacenará la respuesta que nos de el servidor (AKA la información de la página)
respuesta = requests.get(url = "https://es.wikipedia.org/w/api.php", #este es el endpoint para Wikipedia en español, es decir desde donde extraera la info
                         params= {
                              "action": "parse", #Acción que le indica al API qué acción ejecutar. En este caso "analizar"
                              "page": "Época del terrorismo en el Perú", # El nombre de la página de Wikipedia de donde queremos extraer la data
                              "format": "json", #El formato de la respuesta. En este caso es JSON, podría también ser XML u otros
                              "prop": "text", #indica que nos de el contenido en texto versión HTML de la página
                              "redirects": 1 #para evitar que redireccione la pagina
                         })

respuesta #el output nos muestra "Response [200]" que significa que la respuesta fue exitosa.
# Sin embargo, no podemos ver el contenido del objeto. Para ello, debemos hacer un segundo paso, para leer la respuesta que esta en formato json

<Response [200]>

In [ ]:
# La forma más rápida es imprimir el "texto" del objeto respuesta
print(respuesta.text)

{"parse":{"title":"\u00c9poca del terrorismo en el Per\u00fa","pageid":7416800,"redirects":[],"text":{"*":"<div class=\"mw-content-ltr mw-parser-output\" lang=\"es\" dir=\"ltr\"><table style=\"\" class=\"noprint plainlinks ambox ambox-content\">\n<tbody><tr>\n<td class=\"ambox-image\" width=\"40\">\n<span class=\"mw-default-size\" typeof=\"mw:File\"><a href=\"/wiki/Archivo:Commons-emblem-scales.svg\" class=\"mw-file-description\"><img alt=\"\" src=\"//upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/40px-Commons-emblem-scales.svg.png\" decoding=\"async\" width=\"40\" height=\"40\" class=\"mw-file-element\" srcset=\"//upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/60px-Commons-emblem-scales.svg.png 1.5x, //upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/80px-Commons-emblem-scales.svg.png 2x\" data-file-width=\"48\" data-file-height=\"48\" /></a></span></td>\n<td class=\"ambox-text\"><div class=\"ambox-te

In [ ]:
# Una segunda forma es con la función ".json" que permite guardarlo en un nuevo objeto (llamado data) y luego imprimirlo.
# La ventaja de trabajar con formato JSON es que es más amigable para el ojo humano y un formato más universal (máquinas).
# Adicionalmente, con esta segunda forma, hay más elementos que aparecen, por ejemplo las imágenes.
data = respuesta.json()  # La repuesta
print(data)

{'parse': {'title': 'Época del terrorismo en el Perú', 'pageid': 7416800, 'redirects': [], 'text': {'*': '<div class="mw-content-ltr mw-parser-output" lang="es" dir="ltr"><table style="" class="noprint plainlinks ambox ambox-content">\n<tbody><tr>\n<td class="ambox-image" width="40">\n<span class="mw-default-size" typeof="mw:File"><a href="/wiki/Archivo:Commons-emblem-scales.svg" class="mw-file-description"><img alt="" src="//upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/40px-Commons-emblem-scales.svg.png" decoding="async" width="40" height="40" class="mw-file-element" srcset="//upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/60px-Commons-emblem-scales.svg.png 1.5x, //upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/80px-Commons-emblem-scales.svg.png 2x" data-file-width="48" data-file-height="48" /></a></span></td>\n<td class="ambox-text"><div class="ambox-text-div"><strong>Existen desacuerdos sobre la

La solicitud que realizamos a la API nos permitió acceder a la información requerida. Sin embargo, si tuviesemos que trabajar extrayendo una mayor cantidad de data de diferentes páginas, sería importante estandarizar este proceso y automatizarlo. Para ello, definiremos una función que nos permita extraer el contenido de la API solo con una línea de código especificando únicamente el nombre de la página en Wikipedia.

In [ ]:
# Definimos una función llamada 'extraer_contenido' que tenga como único argumento el titulo de la pagina web que requerimos.
def extraer_contenido(titulo):
  # Paso 1: Definimos el URL y los parámetros que queremos utilizar
  url = "https://es.wikipedia.org/w/api.php"
  parametros = {
      "action": "parse",
      "page": titulo,
      "format":"json",
      "prop": "text"
  }

  # Paso 2: La función enviará la solicitud a la API dado los parámetros especificados en el Paso 1
  respuesta = requests.get(url, params=parametros)

  # Paso 3: Leer la respuesta de la API que se encuentra en formato JSON en un dictionario estructurado en Python
  data = respuesta.json()

  # Paso 4: Extraer el código html del formato JSON a un objeto llamado pagina_html y luego imprimirlo
  page_html = data.get("parse", {}).get("text", {}).get("*", "")
  return page_html


In [ ]:
# Ahora que hemos creado nuestra función, podemos aplicarla en solo una linea de código!
extraer_contenido("Época del terrorismo en el Perú")

#Probemoslo en un visualizador de HTML: https://html.onlineviewer.net/
#Copien y peguen el output que figura aquí en esa página y podemos comprobar que se extrajo.

'<div class="mw-content-ltr mw-parser-output" lang="es" dir="ltr"><table style="" class="noprint plainlinks ambox ambox-content">\n<tbody><tr>\n<td class="ambox-image" width="40">\n<span class="mw-default-size" typeof="mw:File"><a href="/wiki/Archivo:Commons-emblem-scales.svg" class="mw-file-description"><img alt="" src="//upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/40px-Commons-emblem-scales.svg.png" decoding="async" width="40" height="40" class="mw-file-element" srcset="//upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/60px-Commons-emblem-scales.svg.png 1.5x, //upload.wikimedia.org/wikipedia/commons/thumb/c/c6/Commons-emblem-scales.svg/80px-Commons-emblem-scales.svg.png 2x" data-file-width="48" data-file-height="48" /></a></span></td>\n<td class="ambox-text"><div class="ambox-text-div"><strong>Existen desacuerdos sobre la <a href="/wiki/Wikipedia:Punto_de_vista_neutral" title="Wikipedia:Punto de vista neutral">neutralidad en

In [ ]:
# Finalmente, ahora que accedimos a la información hay que guardarla.
# Para ello crearemos una función(llamada html_archivo) que guarde el output especificando dos argumentos.
# 1) será el objeto (en Python) con el contenido, y 2) el segundo el nombre del archivo .

def html_archivo(contenido_html, nombre_archivo="page.html"):
    with open(nombre_archivo, "w", encoding="utf-8") as file: #especificamos que guarde con encoding utf-8 que lee caracteres especiales
        file.write(contenido_html)
    print(f"HTML content saved to {nombre_archivo}") #le pedimos a la función que cuando guarde el objeto imprima un mensaje que indique que se guardó exitosamente

# Antes de aplicar la función primero hay que guardar el contenido en un objeto, al que llamaremos CAI_Peru
CAI_Peru = extraer_contenido("Época del terrorismo en el Perú")

# Utilizamos la función que creamos para guardar el objeto en un archivo html
html_archivo(CAI_Peru, "CAI_Peru_Wikipedia.html")

# Ahora podemos comprobar que el archivo se guardo correctamente. Descarguemoslo porque eso lo utilizaremos en la siguiente sesión!

HTML content saved to CAI_Peru_Wikipedia.html


## Parte 2.2 - Ejercicio 1: Extraer historial de revisiones de usuarios



Ahora la idea es que ustedes puedan familiarizarse más con la extracción de información usando la API extrayendo otro tipo de información que la API brinda. <br></br> Wikipedia registra el historial de revisiones que los/as usuarios ejecutan en cada una de las páginas. A cada una de estas modificaciones se les conoce como revisiones y se puede acceder a su historial y visualizar como esta ha cambiado a través del tiempo. <br></br> Cada revisión tiene un identificador único (ID), el nombre del usuario que realizó la modificación, el tipo de modificación que se hizo, entre otros datos relevantes que podríamos utilizar para ver su evolución en el tiempo. <br></br>  Para este ejercicio, se les pide que con la ayuda de la [documentación oficial de la API de Wikipedia](https://www.mediawiki.org/wiki/API:Allrevisions) y herramientas externas (StackOverflow, ChatGpt, etc.) puedan extraer el historial de revisiones para esta página. <br></br> Para entender mejor el historial de revisiones en Wikipedia pueden revisar [siguiente página](https://en.wikipedia.org/wiki/Help:Page_history#) de Wikipedia, lo coloco en la versión en inglés ya que contiene información más detallada que su versión en español


In [ ]:
# En estas líneas de código podrán encontrar la respuesta al ejercicio
# Definimos una función que extraiga el historial de la página de Wikipedia requerida
def extraer_historial_revision(titulo):
 url = "https://es.wikipedia.org/w/api.php"
 parametros = {
       "action": "query", # la acción ya no es analizar, sino "buscar"
       "format": "json",
       "prop": "revisions", # Este parámetro indica que se busquen las revisiones a la página
       "titles": titulo,
       "rvprop": "timestamp|user|comment|ids|size", #las propiedades de cada revisión que pedimos extraer
       "rvlimit": "max", # parámetro que indica que extraiga el número máximo de revisiones por cada solicitud que se haga a la API
       "rvdir": "newer" # el orden de las revisiones será primero la más antigua
 }

#Paso 2: Conectar con la API
 respuesta = requests.get(url, params=parametros)
 data = respuesta.json()

 return data

# Aplicar la función
extraer_historial_revision("Época del terrorismo en el Perú")


In [ ]:
# Segunda parte de la respuesta
# Este output puede llegar a ser un poco desordenado (a nuestra vista humana),
# pero con algunas modificaciones en el codigo se puede convertir en una base de datos tal como la conocemos (matriz)
# Para ello crearemos otra función que además de extraer el historial también convierta el output en una base de datos (matriz).
def extraer_historial_revision_df(titulo):
 url = "https://es.wikipedia.org/w/api.php"
 parametros = {
       "action": "query",
       "format": "json",
       "prop": "revisions",
       "titles": titulo,
       "rvprop": "timestamp|user|comment|ids|size",
       "rvlimit": "max",
       "rvdir": "newer"
 }

 # Paso 2: Conectar con la API
 respuesta = requests.get(url, params=parametros)
 data = respuesta.json()


 # Paso 3: Extraer las listas en format json
 paginas = data.get('query', {}).get('pages', {}) # creamos un diccionario llamado paginas que busca dentro del campo "query" y "pages" del output en formato JSON
 revisiones_por_pagina = [] #creamos una lista vacia para almacenar la información que extraeremos de la respuesta en JSON obtenida

 # Paso 4: Crear un loop que extraiga todas las revisiones por página
 # Cada página posee un identificador único (ID) y tiene la info requerida (información de las revisiones)
 # lo que hace el loop es primero buscar si es que existe el parametro "revisions" (si está presente, significa que hay revisión) en la informacion de la pagina:
 # y si es que está presente, entonces extrae la información ("revisions") y las almacena en la lista vacia "revisiones_por_pagina"
 for id_pagina, info_pagina in paginas.items():
   if "revisions" in info_pagina: # condicional que verifica si es que la pagina tiene revisiones, si cumple con ello, aplica el siguiente código
     revisiones_por_pagina = info_pagina["revisions"] #si la pagina tiene revisiones, añade esas revisiones a la lista vacia "revisiones_por_pagina"


# Paso 5: Crear una lista completa de revisiones
 lista_completa_revisiones = [] # Lista vacia para almacenar las revisiones


# Paso 6: Loop que vaya por cada revision y extraiga cada una de las siguientes variables (revid, timestamp, etc.)
# y lo añada a un nuevo objeto llamado "data_revisiones" que tendra cada revision con sus respectivas variables
 for cada_revision in revisiones_por_pagina:
   data_revisiones = {
       "revid": cada_revision.get("revid"),
       "parentid": cada_revision.get("parentid"),
       "timestamp": cada_revision.get("timestamp"),
       "user": cada_revision.get("user"),
       "size": cada_revision.get("size"),
       "comment": cada_revision.get("comment")
   }
   lista_completa_revisiones.append(data_revisiones) #añadir estas variables a la lista vacia que creamos anteriormente

# Paso 6: Convertir la lista de revisiones en una base de datos (matriz). Para ello usaremos la libreria pandas
 base_datos = pd.DataFrame(lista_completa_revisiones)


 return base_datos

# Debemos importar la libreria pandas para la ultima parte de la funcion
import pandas as pd


# Aplicar la función creada
extraer_historial_revision_df("Época del terrorismo en el Perú")


#El resultado son 500 revisiones distintas...


revid  parentid             timestamp  \
0    12620573         0  2007-11-06T18:00:19Z   
1    12620598  12620573  2007-11-06T18:01:08Z   
2    12620620  12620598  2007-11-06T18:01:39Z   
3    12620862  12620620  2007-11-06T18:09:45Z   
4    12620891  12620862  2007-11-06T18:10:43Z   
..        ...       ...                   ...   
495  88185134  88185007  2016-01-03T04:11:38Z   
496  88190736  88185134  2016-01-03T13:37:48Z   
497  88312615  88190736  2016-01-08T20:40:48Z   
498  88328051  88312615  2016-01-09T14:30:48Z   
499  88539656  88328051  2016-01-18T19:47:33Z   

                                        user   size  \
0                                  Jorgejhms   1627   
1                                  Jorgejhms   1626   
2                                  Jorgejhms   1624   
3                                  Jorgejhms   2093   
4                                  Jorgejhms   2166   
..                                       ...    ...   
495  2001:1388:3682:76D1:2563:EEC6:EB2F:484F  37461   
496                                 Mansoncc  37240   
497                           190.187.95.138  37091   
498                                 Mansoncc  37240   
499                               Diánmondin  37296   

                                               comment  
0    Página nueva: {{traducción|ingles|Internal con...  
1                                           corrección  
2                         /* Referencias */ corrección  
3                                           ampliación  
4              /* Referencias */ sección vease también  
..                                                 ...  
495                                                     
496  Revertidos los cambios de [[Special:Contributi...  
497                 /* Gobierno de Alberto Fujimori */  
498  Revertidos los cambios de [[Special:Contributi...  
499                                                     

[500 rows x 6 columns]

# Parte 3: Extraer información de la web sin el uso de APIs

El objetivo de esta sección es familiarizarse con la extracción de diferentes elementos de una página web. Para ello, hemos elegido la siguiente noticia del periódio "Correo": [Ica: caen tres extranjeros que integrarían “Los Gallegos de Ica”, dedicados al peligroso “gota a gota”](https://diariocorreo.pe/edicion/ica/ica-caen-integrantes-de-la-los-gallegos-de-ica-noticia/).

La idea es familiarizarnos con extraer diferentes elementos, para posteriormente, aplicando la  misma lógica,  extraer noticias que nos ayuden a responder el objetivo que nos planteamos: "Cómo se representan a los migrantes de origen Venezolano en el Perú".

In [ ]:
# Instalaremos la libreria "Beautiful Soup" que nos permite navegar documentos en formato HTML y XML.
!pip install bs4
#Recuerden que una ves instalada ya no se necesita instalarla nuevamente

In [ ]:
# Ahora la importaremos para poder utilizarla. Recuerden que este paso se tiene que hacer todas las veces que iniciamos el programa.
# es como cargar una libreria en R.
from bs4 import BeautifulSoup

import requests # por si acaso

## 3.1. Extraer el titular de la noticia


Antes de extraer la información debemos identificar en qué parte del documento HTML se encuentra el elemento que queremos extraer. Comencemos, entonces, extrayendo el titular.

Primero debemos dirigirnos al [siguiente enlace](https://diariocorreo.pe/edicion/ica/ica-caen-integrantes-de-la-los-gallegos-de-ica-noticia/), seleccionar manualmente el título, colocar click derecho y seleccionar "Inspeccionar".

Automáticamente se abrirá una panatalla de color negra en la franja derecha mostrando el código HTML y especificamente la ubicación de este elemento:

> < h1 itemprop="name" class="sht__title">Ica: caen tres extranjeros que integrarían  “Los Gallegos de Ica”, dedicados al peligroso “gota a gota”</ h1>

El elemento que queremos tiene el tag de < h1> el cual indica que es un "encabezado  nivel 1" (El que tiene mayor jerarquía existe hasta h6) y cuenta con dos atributos:
1. itemprop = "name" (parte de la metadata)
2. class = "sht__title" (de CSS, indica un tipo de estilo)

Luego se muestra el contenido en texto que es al que queremos acceder.Una vez identificado el elemento a extraer, podemos ir al programa e indicarle que queremos que extraiga ese elemento en específico.

In [ ]:
# Paso 1: Definimos el enlace web de donde queremos extraer la información (noticia) en un objeto llamado "url"
url = 'https://diariocorreo.pe/edicion/ica/ica-caen-integrantes-de-la-los-gallegos-de-ica-noticia/'

In [ ]:
# Paso 2: Utilizamos la libreria request para poder conectar con dicho URL (acordemonos de cómo lo utilizamos para conectarnos con la API de Wikipedia)
# Para ello, aplicamos el comando requests.get para conectar con el contenido de ese URL y extraer su información en formato HTML.
# La respuesta del servidor se almacena en el objeto "respuesta_servidor"

respuesta_servidor = requests.get(url) # el objeto "request" contiene el contenido HTML de Wikipedia

In [ ]:
# Paso 3: Utilizamos el comando de BeautifulSoup para pedir que se lea la rspuesta del servidor que esta en formato HTML

noticia_html = BeautifulSoup(respuesta_servidor.text, # obtiene el código (en texto) de la respuesta del servidor (como en el caso del API, recordemos)
                             'html.parser') # este parametro indica que el formato del documento es html

# El output que se guarda en el objeto "soup" es un objeto que permite navegar el formato HTML con esta libreria y aplicar todas sus funcionalidades.

In [ ]:
# Comprobemos que el objeto existe...

noticia_html

# El output se ve "raro" porque se encuentra en un formato específico para que la librería BeautifulSoup lo pueda "leer".
# Ahora la idea es utilizar los comandos de la libreria BeautifulSoup para "navegar" el objeto.

In [ ]:
# Busquemos el titulo por su tag principal que es "h1" (Encabezado nivel 1 AKA titulo principal)

noticia_html.h1

# El resultado es exactamente el titular de la noticia

<h1 class="sht__title" itemprop="name">Ica: caen tres extranjeros que integrarían  “Los Gallegos de Ica”, dedicados al peligroso “gota a gota”</h1>

Pero en un mismo documento podrían haber varios elementos que tengan el tag "h1". Si bien esto es raro -ya que usualmente solo hay un título- esta situación se podría presentar cuando queramos extraer otros elementos, como por ejemplo cada uno de los párrafos del texto del artículo. Por ello, otra forma de extraer el titular es mediante la búsqueda de uno de sus atributos. Por ejemplo el atributo "class"




In [ ]:
# Utilizamos el comando find_all, para ubicar todos los elementos que del atributo "class" tengan el valor "sht_title".
noticia_html.find_all(class_ = "sht__title")

# El resultado es el mismo

[<h1 class="sht__title" itemprop="name">Ica: caen tres extranjeros que integrarían  “Los Gallegos de Ica”, dedicados al peligroso “gota a gota”</h1>]

## 3.2. Extraer el resumen de la noticia




Ahora ubiquemos cómo extraer otros elementos tales como la sumilla de la noticia. Para ello iremos a la página web nuevamente y ubicaremos en qué lugar del "árbol" se encuentra.



> < h2 itemprop="name" class="sht__summary ">Los sujetos fueron intervenidos en flagrancia, lo que permitió su arresto inmediato.</ h2>


Para este caso, la sumilla se encuentra con el tag "h2" o en la clase "sht__summary"



In [ ]:
# Extraigamos la sumilla utilizando "h2"
noticia_html.h2

# El output muestra algo que no era lo que nosotros vimos, ¿Por qué? Pues por el simple hecho de que hay varios elementos que tienen el tag h2

<h2 class="st-social__txt oflow-h uppercase" itemprop="name"><a class="st-social__link oflow-h" href="/edicion/ica/" itemprop="url">Ica</a></h2>

In [ ]:
# Aquí podemos corroborar esto:
noticia_html.find_all("h2")

# El resultado muestra todos los elementos con tag "h2", dentro de los cuáles la sumilla es el número 2. Por ello debemos ser más específicos

[<h2 class="st-social__txt oflow-h uppercase" itemprop="name"><a class="st-social__link oflow-h" href="/edicion/ica/" itemprop="url">Ica</a></h2>,
 <h2 class="sht__summary" itemprop="name">Los sujetos fueron intervenidos en flagrancia, lo que permitió su arresto inmediato.</h2>,
 <h2 class="story-contents__header"><b>Banda capturada</b></h2>,
 <h2 class="story-contents__header"><b>VIDEO RECOMENDADO</b></h2>,
 <h2 class="link-site__title-link oflow-h" itemprop="name">Horóscopo HOY sábado 21 de setiembre con las predicciones de Carmen Briceño según tu signo</h2>,
 <h2 class="link-site__title-link oflow-h" itemprop="name">Ica: delincuentes roban transformador y dejan a 200 familias sin electricidad en Los Molinos</h2>,
 <h2 class="link-site__title-link oflow-h" itemprop="name">Trujillo: Delincuentes disparan en la pierna a hombre para robarle S/2 mil </h2>]

In [ ]:
noticia_html.find_all(class_ = "sht__summary")

# Ahora sí obtuvimos lo que queríamos...

[<h2 class="sht__summary" itemprop="name">Los sujetos fueron intervenidos en flagrancia, lo que permitió su arresto inmediato.</h2>]

**OJO: Cuando hemos copiado el código directamente del html este aparece como "sht__summary " (con un espacio entre la última letra y las dos comillas). Sin embargo, si es que ustedes corren el código con ese espacio no va a figurar nada. Deben borrarlo para que aparezca. Siempre recuerden revisar bien estos pequeños detalles.**

## 3.3. Ejercicio 2: Extraer el cuerpo de la noticia

Ahora les toca a ustedes. La idea es que puedan extraer todo el cuerpo de la noticia. Para ello deberán ubicar cuáles son los elementos a extraer y luego utilizar la misma lógica para hacerlo

In [ ]:
# Respuesta

noticia_html.find_all(class_ = "story-contents__font-paragraph")


[<p class="story-contents__font-paragraph" itemprop="description"><b>En una operación conjunta entre la Dirección General de Inteligencia del Ministerio del Interior (Digimin) y el Grupo Terna del Frente Policial de </b><a href="https://diariocorreo.pe/edicion/ica/ica-atropellan-a-joven-ciclista-frente-a-la-universidad-san-luis-gonzaga-noticia/" itemprop="url" target="_blank"><b>Ica</b></a><b>, se logró desarticular a la </b><a href="https://diariocorreo.pe/edicion/ica/ica-sujeto-es-sentenciado-a-nueve-meses-de-prision-preventiva-por-integrar-banda-que-enganaba-a-transportistas-en-semaforos-noticia/" itemprop="url" target="_blank"><b>banda </b></a><b>delictiva internacional denominada “Los Gallegos de Ica” el pasado 6 de septiembre.</b></p>,
 <p class="story-contents__font-paragraph" itemprop="description">La intervención tuvo lugar entre las 19:00 y 22:00 horas en una vivienda ubicada en la urbanización Las Casuarinas, 5ta etapa, en la calle Ucayali B-23 de la ciudad de Ica.</p>,
 <p 

## 3.4. Almacenar diferentes párrafos extraídos

La gran pregunta es, ¿cómo puedo guardar esta data? En este momento lo que tenemos es un objeto creado que se llama

```
pagina_html
```
El cual contiene toda la inforamción de la página en HTML. En este sentido, tenemos dos formas de guardar la información:
1. Guardando el objeto como tal (con toda la información)
2. Guardar solo los elementos que nos interesan en un objeto como una base de datos


In [ ]:
# Forma 1: Guardar el objeto HTML

with open("noticia.html", "w", encoding="utf-8") as file:
    file.write(noticia_html.prettify())

# Revisamos nuestra "carpetita" en Google Collab y se ha guardado como archivo html que podemos descargarlo a nuestro disco local

In [ ]:
# Forma 2: Guardar solo los elementos requeridos en formato json
import json

#Guardar el contenido en parrafos_noticia
parrafos_noticia = noticia_html.find_all(class_ = "story-contents__font-paragraph")

# Extraer cada párrafo en una lista
parrafos_lista = [p.get_text() for p in parrafos_noticia]

# Guardar la lista en un archivo JSON
with open("parrafos.json", "w", encoding="utf-8") as file:
    json.dump(parrafos_lista, file, ensure_ascii=False, indent=4)

## 3.5. Extraer varios titulares (loop)

Con lo aprendido previamente, el reto es  extraer todos los titulares de las últimas 3 páginas, de modo que podamos hacer una base de datos más grande de la cual extrar información y responder a la pregunta que tenemos.

Para ello, primero debemos definir una palabra de búsqueda que nos permita acceder a las noticias que refieran a "migrantes venezolanos". Para ello utilizaremos la palabra "venezolano".

Si colocamos esta palabra en el motor de búsqueda de la página web del Correo, entonces podemos entender la lógica detrás del motor de búsqueda y modificarlo para poder generar una base de datos que nos permita extraer las noticias de las últimas 5 páginas.

> https://diariocorreo.pe/buscar/venezolano/todas/descendiente/1/

Cada página muestra 30 noticias, por lo que si extraemos las primeras 5 páginas, tendremos un aproximado de 150 titulares. Para ello, utilizaremos la siguiente lógica:

1. Crear una base de datos con los 5 enlaces (loop)
2. Identificar el titular en el árbol HTML
3. Diseñar una función de extracción
5. Guardar el contenido en una base de datos (matriz)



In [ ]:
# Paso 1: Crear una base de datos con los enlaces para las primeras 5 noticias

enlace_base= 'https://diariocorreo.pe/buscar/venezolano/todas/descendiente/' #definimos en enlace base

pagina_inicial = 1
pagina_final = 5

# Creamos una lista vacia para almacenar todos estos enlaces
enlaces_correo = []

# Loop para agregar el numero de cada página
for numero_pagina in range(pagina_inicial, pagina_final + 1): # por cada pagina en el rango de 1 al 5 haz lo siguiente
  url = f'{enlace_base}{numero_pagina}/' #junta el 'enlace_base' con 'el numero de la pagina actual' y añadele "/". guardalo en un objeto llamado "url"
  enlaces_correo.append(url) # Añade el url anterior a la lista vacia 'enlaces_correo'

# Mostramos nuestra lista
print(enlaces_correo)

['https://diariocorreo.pe/buscar/venezolano/todas/descendiente/1/', 'https://diariocorreo.pe/buscar/venezolano/todas/descendiente/2/', 'https://diariocorreo.pe/buscar/venezolano/todas/descendiente/3/', 'https://diariocorreo.pe/buscar/venezolano/todas/descendiente/4/', 'https://diariocorreo.pe/buscar/venezolano/todas/descendiente/5/']


**Paso 2: Identificar el titular en el código HTML**

Ahora que ya tenemos nuestra base de datos con las 5 primeras páginas de búsqueda de noticias, es momento de identificar el tag que contiene el titular de cada noticia:

> < a itemprop="url" class="story-item__title block overflow-hidden primary-font line-h-xs mt-10" href="/edicion/huancayo/huancayo-ocultan-mas-de-2-toneladas-de-acido-sulfurico-en-ladrillos-y-cemento-noticia/">Huancayo: Ocultan en camión más de 2 toneladas de ácido sulfúrico en ladrillos y cemento</a >

Con este código, identificamos que para acceder al titular, debemos acceder por medio del atributo "class" con el siguiente valor: "class="story-item__title block overflow-hidden primary-font line-h-xs mt-10"

In [ ]:
# Paso 3: Definir una función de extracción

# Definir una base de datos vacias para almacenar los titutares
correo_df = pd.DataFrame(columns=['titular'])

# Diseñar una función que se diriga a cada url y extraiga los titulares y luego los añada dentro de una base de datos
def extraer_titular(lista_url):
  global correo_df
  for url in lista_url:
    respuesta_servidor = requests.get(url) # conectar con el servidor
    pagina_html = BeautifulSoup(respuesta_servidor.text, 'html.parser') # transformarlo en objeto BeautifulSoup
    titulares = pagina_html.find_all('a', class_ = 'story-item__title block overflow-hidden primary-font line-h-xs mt-10') #Extrer el tiutlar

    #Añadir en una base de datos
    for titular in titulares:
      correo_df = pd.concat([correo_df, pd.DataFrame({'titular': [titular.get_text()]})], ignore_index=True)

    # En caso sea una base mas grande se puede añadir una pausa entre 1 y 3 segundos para no dañar el servidor
    #time.sleep(random.uniform(1, 3)) # se necesita libreria 'time' y 'random'

In [ ]:
# Paso 4: Aplicar la función
extraer_titular(enlaces_correo)

print(correo_df)

                                               titular
0    Edmundo González Urrutia aclara que no fue coa...
1    Venezuela pedirá arresto de Milei por caso de ...
2    Palpa: extranjero atacó con un fierro a su com...
3    Gobierno de Venezuela arrestó a un cuarto esta...
4    Cuerpo hallado en el mar de Arequipa, sería de...
..                                                 ...
145  TODA TIRANÍA TIENE IDEOLOGÍA, columna de Martí...
146  OEA pedirá a la Corte Penal Internacional el a...
147  Trujillo: Empresario dispara contra presuntos ...
148  Intervienen a mujer que amenazaba con un cuchi...
149  Nicolás Maduro sobre Edmundo González y María ...

[150 rows x 1 columns]


In [ ]:
# Paso 5: Guardar la base de datos
correo_df.to_csv('correo_df.csv', index=False)